thunder/dynamo/splitter.py

```diff
@@ -0,0 +1,166 @@
+        class GraphModule(torch.nn.Module):
+            def forward(self, l_x_: "f32[2]"):
+                inductor_1 = self.inductor_1(l_x_)
+                thunder_2 = self.thunder_2(l_x_, inductor_1);  l_x_ = inductor_1 = None
+                return (thunder_2,)
+
+            class inductor_1(torch.nn.Module):  # Partition for inductor
+                def forward(self, l_x_: "f32[2]"):
+                    y: "f32[2]" = torch.sinc(l_x_);  l_x_ = None
+                    return y
+
+            class thunder_2(torch.nn.Module):  # Partition for thunder
+                def forward(self, l_x_: "f32[2]", y: "f32[2]"):
+                    matmul: "f32[]" = torch.matmul(l_x_, y);  l_x_ = y = None
+                    return matmul
+    """
+    # The callback below is called for every node in the graph.
+    # It returns an `int` denoting the parition where the node should be placed.
+    # We want to partition the graph into contiguous regions (with one or more operations)
+    # into thunder supported or unsupported region.
+    # `prev_value` is used to determine if we are still in same region (i.e. supported region or unsupported region).
+    # `partition_cnt` is bumped everytime we change the region i.e. flip from supported to unsupported or from unsupported to supported.
+    # `supported_partitions` is used to track the thunder supported partitions.
+    prev_value = None
+    partition_cnt = 0
+    supported_partitions: set[int] = set()
+    split_reasons: list[SplitReason] = []
+
+    nodes_in_unsupported_ctx_regions = get_nodes_in_unsupported_ctx_regions(gm)
+
+    def callback(node) -> int:
+        nonlocal prev_value, partition_cnt, split_reasons, supported_partitions
+
+        assert node.op not in (
+            "placeholder",
+            "get_attr",
+            "output",
+        ), f"fx.split_module should have only passed node.op=call_* but received {node.op}"
+
+        if node in nodes_in_unsupported_ctx_regions:
+            # If node was in unsupported ctx region like `autocast`,
+            # even though the operation maybe supported, we pass it to `torch.compile`
+            # as `thunder` doesn't correctly work with these.
+            is_thunder_supported = False
+            split_reason = SplitReason(
+                SplitReasonType.UNSUPPORTED_NODE,
+                info=f"node with name: {node.name} and target: {node.target} is not supported probably because it is in unsupported context.",
+            )
+            split_reasons.append(split_reason)
+        else:
+            is_thunder_supported, split_reason = is_node_supported_by_thunder(node)
+            if split_reason is not None:
+                split_reasons.append(split_reason)
+
+        if prev_value == is_thunder_supported:  # We are in the same region.
+            return partition_cnt
+
+        # There is a flip. Either from supported to unsupported or unsupported to supported.
+        prev_value = is_thunder_supported
```

Comment:
It might be helpful to use `groupby` (or another itertools utility): 

https://docs.python.org/3/library/itertools.html#itertools.groupby